### PR TITLE
Port Primitive Types Tests

### PR DESCRIPTION
**Pull request by WanjohiSammy**
### Issues

*This pull request fixes #xxx.*

### Description

Port QueryOption Tests from [`WCF`](https://learn.microsoft.com/en-us/dotnet/framework/wcf/whats-wcf) to use `AspNetCoreOData`

- [test/EndToEndTests/Tests/Client/Build.Desktop/PrimitiveTypesTests](https://github.com/OData/odata.net/tree/main/test/EndToEndTests/Tests/Client/Build.Desktop/PrimitiveTypesTests)
- [test/EndToEndTests/Tests/Client/Build.Desktop/KeyAsSegmentTests/PrimitiveKeyValuesTests](https://github.com/OData/odata.net/blob/main/test/EndToEndTests/Tests/Client/Build.Desktop/KeyAsSegmentTests/PrimitiveKeyValuesTests.cs)

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/PrimitiveKeysValuesTests.cs [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+        {
+            return float.IsNaN(key);
+        }
+
+        private static bool IsNotSupportedKey(double key)
+        {
+            return double.IsNaN(key);
+        }
+
+        private void ResetDefaultDataSource()
```
Comment (xuzhg) [inline test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/PrimitiveKeysValuesTests.cs:238]: any HttpRequest test to call this reset method?

**Review by xuzhg — APPROVED** [review]

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/PrimitiveKeysValuesTests.cs (hunk 1 above)
Comment (WanjohiSammy) [reply test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/PrimitiveKeysValuesTests.cs:238]: This method is called from the constructor of `PrimitiveKeysValuesTests`:

```cs
 public PrimitiveKeysValuesTests(TestWebApplicationFactory<PrimitiveKeysValuesTests.TestsStartup> fixture)
     : base(fixture)
 {
     // The rest of code
     ResetDefaultDataSource(); // here
 }
```

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/PrimitiveKeysValuesTests.cs [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+            };
+
+            _model = PrimitiveKeyValuesEdmModel.GetEdmModel();
+            ResetDefaultDataSource();
+        }
+
+        [Fact]
+        public void BinaryTest()
+        {
+            foreach (var entry in _context.EdmBinarySet)
```
Comment (habbes) [inline test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/PrimitiveKeysValuesTests.cs:69]: Instead of having a loop inside the test, could make this a `[Theory]` with `[TheoryDataSet]` instead of a `[Fact]` and use the data set? That way the test could be executed separately for each input and it would be easier to trace failing tests.

Comment (WanjohiSammy) [reply test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/PrimitiveKeysValuesTests.cs]: Thanks for noting this. I have used `[Theory]` with `TheoryData`

File: test/EndToEndTests/Common/Microsoft.OData.Client.E2E.TestCommon/Helpers/ODataMessageReaderTestsHelper.cs [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+                }
+                Assert.Equal(ODataReaderState.Completed, reader.State);
+            }
+
+        }
+
+        return entries;
+    }
+
+    public async Task<(List<ODataResource>, List<ODataResourceSet>)> QueryResourceAndResourceSetsAsync(string queryText, string mimeType)
```
Comment (habbes) [inline test/EndToEndTests/Common/Microsoft.OData.Client.E2E.TestCommon/Helpers/ODataMessageReaderTestsHelper.cs:110]: Would be good to have doc comments for this method as well. Also it would help understand this method (and other methods in this class) if we could see examples of queries or URLs it expects.

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/DurationTests.cs [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+
+        _model = DefaultEdmModel.GetEdmModel();
+        ResetDefaultDataSource();
+    }
+
+    public static IEnumerable<object[]> MimeTypesData
+    {
+        get
+        {
+            yield return new object[] { MimeTypes.ApplicationJson + MimeTypes.ODataParameterFullMetadata };
```
Comment (habbes) [inline test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/DurationTests.cs:63]: Are these mime types `const` strings? If they are constants, then I would suggest using `InlineData` instead of `MemberData`.

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/DurationTests.cs [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+            yield return new object[] { MimeTypes.ApplicationJson + MimeTypes.ODataParameterMinimalMetadata };
+            yield return new object[] { MimeTypes.ApplicationJson + MimeTypes.ODataParameterNoMetadata };
+        }
+    }
+
+    #region Entity Set
+
+    [Theory]
+    [MemberData(nameof(MimeTypesData))]
```
Comment (habbes) [inline test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/DurationTests.cs:72]: If these mime types are constants, then I'd suggest using inline data as so:

```suggestion
    [InlineData(MimeTypes.ApplicationJson + ODataParameterFullMetadata)]
    [InlineData(MimeTypes.ApplicationJson + MimeTypes.ODataParameterMinimalMetadata)]
    [InlineData(MimeTypes.ApplicationJson + MimeTypes.ODataParameterNoMetadata)]
```

Why? This makes the tests easier to read. It's easier to know what is being tested without having to search for the definition of the member data collection. We should strive to keep the tests self-contained when feasible, so that we don't have to look at other source code to understand what the test is trying to achieve.

If the fact that these mime types are constant is not important to the feature, we could even go further and write them out as string literals.

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/DurationTests.cs [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+    public async Task QueryEntitySetTest(string mimeType)
+    {
+        // Arrange & Act
+        var queryText = "Customers";
+        var entries = await this.TestsHelper.QueryResourceSetsAsync(queryText, mimeType);
+
+        // Assert
+        if (!mimeType.Contains(MimeTypes.ODataParameterNoMetadata))
```
Comment (habbes) [inline test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/DurationTests.cs:80]: Why do we have an if statement here? If this test does not make sense for ODataParameterNoMetadata, then we should not include it in the test. Again, using InlineData would make it easier to include only the relevant mime types.

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/DurationTests.cs [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+        {
+            foreach (var entry in entries)
+            {
+                if (entry.TypeName.EndsWith("Customer"))
+                {
+                    var timeBetweenLastTwoOrdersProperty = entry.Properties.Single(p => p.Name == "TimeBetweenLastTwoOrders") as ODataProperty;
+                    Assert.NotNull(timeBetweenLastTwoOrdersProperty);
+                    Assert.NotNull(timeBetweenLastTwoOrdersProperty.Value);
+                }
+            }
```
Comment (habbes) [inline test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/DurationTests.cs:90]: This test will succeed if the return collection is empty or has no entries with a type name that ends with "Customer". This means that we cannot guarantee that the relevant code path is tested when this test runs. I think we should avoid such conditional tests. We should have the confidence that if this test runs and passes, then the associated behavior is correctly implemented.  If not, how can we make this test more specific and deterministic. If we expect that query to return a non-empty array, then we should test for that. If we expect there to be customer entries, we should retrieve them and ensure that set is non empty, or has the expected number of elements. You can look for a way to refactor this such that there's guarantee that when this test runs, the behaviour is actually tested. When possible, let's avoid complex logic such conditional and loops in tests.

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/DurationTests.cs [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+        // Arrange & Act
+        var queryText = "Customers(1)";
+        var entries = await this.TestsHelper.QueryResourceEntriesAsync(queryText, mimeType);
+
+        // Assert
+        if (!mimeType.Contains(MimeTypes.ODataParameterNoMetadata))
+        {
+            foreach (var entry in entries)
+            {
+                if (entry.TypeName.EndsWith("Customer"))
```
Comment (habbes) [inline test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/DurationTests.cs:111]: See https://github.com/OData/odata.net/pull/3160/files#r1914339945

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/DurationTests.cs [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+    [Theory]
+    [MemberData(nameof(MimeTypesData))]
+    public async Task QueryEntityPropertyTest(string mimeType)
+    {
+        // Arrange & Act
+        var queryText = "Customers(1)/TimeBetweenLastTwoOrders";
+        var property = await this.TestsHelper.QueryPropertyAsync(queryText, mimeType);
+
+        // Assert
+        if (!mimeType.Contains(MimeTypes.ODataParameterNoMetadata))
```
Comment (habbes) [inline test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/DurationTests.cs:134]: See https://github.com/OData/odata.net/pull/3160/files#r1914339945

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/DurationTests.cs [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+        var queryText = "Orders(7)/CustomerForOrder";
+        var entries = await this.TestsHelper.QueryResourceEntriesAsync(queryText, mimeType);
+
+        // Assert
+        if (!mimeType.Contains(MimeTypes.ODataParameterNoMetadata))
+        {
+            foreach (var entry in entries)
+            {
+                if (entry.TypeName.EndsWith("Customer"))
+                {
```
Comment (habbes) [inline test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/DurationTests.cs:159]: See https://github.com/OData/odata.net/pull/3160/files#r1914339945

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/PrimitiveKeysValuesTests.cs [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+        [Theory]
+        [MemberData(nameof(GetEdmDoubleSet))]
+        public void DoubleTest(EdmDouble entry)
+        {
+            // Arrange & Act
+            var query = _context.CreateQuery<EdmDouble>("EdmDoubleSet").Where(e => e.Id.Equals(entry.Id));
+            var queryResult = query.ToArray();
+
+            // Assert
+            //Expected a single result for key value {0}, entry.Id.ToString(CultureInfo.InvariantCulture)
```
Comment (habbes) [inline test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/PrimitiveKeysValuesTests.cs:123]: Maybe this comment could be used as the custom message of the assertion?

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/PrimitiveKeysValuesTests.cs [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+            {
+                HttpClientFactory = HttpClientFactory
+            };
+
+            _model = PrimitiveKeyValuesEdmModel.GetEdmModel();
+            ResetDefaultDataSource();
+        }
+
+        [Theory]
+        [MemberData(nameof(GetEdmBinarySet))]
```
Comment (habbes) [inline test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/PrimitiveKeysValuesTests.cs:67]: In this case, I also think there would be value in spelling out all the keys we want to test with using `InlineData`. We could pass a primtive value to the attribute, then in the test, construct the `EdmBinary` instance. I see that advantage with your approach is that the inputs here are the same keys used in the context, and so if the changing it in on place will automatically update it here. But I'm not sure if that beats the readability of clearly seeing what's being tested, which would also make it easy to spot if we're missing any edge cases. A bit data or code duplication is not too bad in tests.

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/PrimitiveKeysValuesTests.cs [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+            var queryResult = query.ToArray();
+
+            // Assert
+            //Expected a single result for key value {0}, entry.Id.ToString()
+            Assert.Single(queryResult);
+        }
+
+        #region TheoryData
+
+        public static TheoryData<T> GetTheoryData<T>(Func<Container, IEnumerable<T>> dataSelector)
```
Comment (habbes) [inline test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/PrimitiveKeysValuesTests.cs:207]:
```suggestion
        private static TheoryData<T> GetTheoryData<T>(Func<Container, IEnumerable<T>> dataSelector)

```

Comment (habbes) [reply test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/PrimitiveKeysValuesTests.cs:207]: But do we really need this method. The logic seems complex. Why do we have `IsNotSupportedKey` getting called here? I think we should write the literal keys in `InlineData` to avoid such "magic" in tests.

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/PrimitiveValueFormatTest.cs [hunk 14]
```diff
@@ -0,0 +1,476 @@
...
+        {
+            yield return new object[] { MimeTypes.ApplicationJson + MimeTypes.ODataParameterFullMetadata };
+            yield return new object[] { MimeTypes.ApplicationJson + MimeTypes.ODataParameterFullMetadata + MimeTypes.ODataParameterIEEE754Compatible };
+        }
+    }
+
+    public string DataPattern = "(" + @"[\+-]?\d\.?\d*E?[\+-]?\d*" + "|" + @"INF|-INF|NaN" + ")";
+
+    [Theory]
+    [MemberData(nameof(MimeTypesData))]
```
Comment (habbes) [inline test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/PrimitiveValueFormatTest.cs:74]: See https://github.com/OData/odata.net/pull/3160/files#r1914339945

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/PrimitiveValueFormatTest.cs [hunk 15]
```diff
@@ -0,0 +1,476 @@
...
+        var (resourceEntries, resourceSetEntries) = await TestsHelper.QueryResourceAndResourceSetsAsync(skipTokenQuery, mimeType);
+
+        var pattern = string.Concat(skipTokenQuery.AsSpan(0, skipTokenQuery.IndexOf('?')), @"\(", DataPattern, @"\)$");
+        var rgx = new Regex(pattern, RegexOptions.IgnoreCase);
+
+        foreach (var entry in resourceEntries)
+        {
+            Assert.True(rgx.Match(entry.Id.ToString()).Success);
+
+            if (mimeType.Equals(MimeTypes.ApplicationJson + MimeTypes.ODataParameterFullMetadata))
```
Comment (habbes) [inline test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/PrimitiveValueFormatTest.cs:445]: See https://github.com/OData/odata.net/pull/3160/files#r1914339945

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/PrimitiveValueFormatTest.cs [hunk 16]
```diff
@@ -0,0 +1,476 @@
...
+
+    [Theory]
+    [MemberData(nameof(MimeTypesData))]
+    public async Task FloatWithoutSuffixAsKeyInURLTest(string mimeType)
+    {
+        // Arrange
+        var expectedKeySegment = "EdmSingleSet(1)";
+
+        // Act & Assert
+        await PrimitiveValueAsKeyInURLAsync("EdmSingleSet(1.0)", mimeType, expectedKeySegment);
```
Comment (habbes) [inline test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/PrimitiveValueFormatTest.cs:104]: It's not clear what this method is doing. Is it possible to write the relevant "act" and "assert" portions directly?

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/DurationTests.cs (hunk 4 above)
Comment (WanjohiSammy) [reply test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/DurationTests.cs]: I have created them as constants and passed them using InlineData

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/DurationTests.cs (hunk 6 above)
Comment (WanjohiSammy) [reply test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/DurationTests.cs]: Resolved this

File: test/EndToEndTests/Common/Microsoft.OData.Client.E2E.TestCommon/Helpers/ODataMessageReaderTestsHelper.cs (hunk 3 above)
Comment (WanjohiSammy) [reply test/EndToEndTests/Common/Microsoft.OData.Client.E2E.TestCommon/Helpers/ODataMessageReaderTestsHelper.cs]: Added the doc string

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/DurationTests.cs (hunk 5 above)
Comment (WanjohiSammy) [reply test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/DurationTests.cs]: I have created them as constants and passed them using InlineData as recommended

File: test/EndToEndTests/Common/Microsoft.OData.Client.E2E.TestCommon/Helpers/ODataMessageReaderTestsHelper.cs (hunk 3 above)
Comment (WanjohiSammy) [reply test/EndToEndTests/Common/Microsoft.OData.Client.E2E.TestCommon/Helpers/ODataMessageReaderTestsHelper.cs]: I have added the doc string

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/DurationTests.cs (hunk 7 above)
Comment (WanjohiSammy) [reply test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/DurationTests.cs]: This was a list of items, that's why I had to loop and ensure the value is not null. I have rectified this to below:
```cs
// Arrange & Act
var queryText = "Customers";
var entries = await this.TestsHelper.QueryResourceSetsAsync(queryText, mimeType);

entries = entries.Where(e => e.TypeName.EndsWith("Customer")).ToList();

// Assert
var timeBetweenLastTwoOrdersProperty = entries[0].Properties.Single(p => p.Name == "TimeBetweenLastTwoOrders") as ODataProperty;
Assert.NotNull(timeBetweenLastTwoOrdersProperty);
Assert.Equal(new TimeSpan(1), timeBetweenLastTwoOrdersProperty.Value);

timeBetweenLastTwoOrdersProperty = entries[1].Properties.Single(p => p.Name == "TimeBetweenLastTwoOrders") as ODataProperty;
Assert.NotNull(timeBetweenLastTwoOrdersProperty);
Assert.Equal(new TimeSpan(2), timeBetweenLastTwoOrdersProperty.Value);
```

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/DurationTests.cs (hunk 9 above)
Comment (WanjohiSammy) [reply test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/DurationTests.cs]: Resolved this.

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/PrimitiveKeysValuesTests.cs (hunk 12 above)
Comment (WanjohiSammy) [reply test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PrimitiveTypesTests/Tests/PrimitiveKeysValuesTests.cs]: I have explicitly passed the IDs using `InlineData` instead of using `MemberData` like below:

```cs
[Theory]
[InlineData(new Byte[] { }, "binary''")]
[InlineData(new Byte[] { 1 }, "binary'AQ%3D%3D'")]
[InlineData(new Byte[] { 2, 3, 4 }, "binary'AgME'")]
public void BinaryTest(byte[] entryId, string expectedIdInUrl)
...
```

```cs
[Theory]
[InlineData(true, "true")]
[InlineData(false, "false")]
public void BooleanTest(bool entryId, string expectedIdInUrl)
...
```

```cs
[Theory]
[InlineData(0, "0.0")]
[InlineData(-1, "-1")]
[InlineData(1, "1.0")]
[InlineData(Double.MaxValue, "1.7976931348623157E%2B308")]
[InlineData(Double.MinValue, "-1.7976931348623157E%2B308")]
[InlineData(Double.Epsilon, "5E-324")]
[InlineData(Double.NegativeInfinity, "-INF")]
[InlineData(Double.PositiveInfinity, "INF")]
public void DoubleTest(double entryId, string expectedIdInUrl)
...
```

**Review by habbes — APPROVED** [review]
